safekeeper/src/wal_backup.rs

```diff
@@ -0,0 +1,267 @@
+use utils::{
+    lsn::Lsn,
+    zid::{ZTenantTimelineId},
+};
+
+use tracing::*;
+
+use crate::config::{RemoteStorageConfig, RemoteStorageKind};
+use crate::remote_storage::{NoopStorage, RemoteStorage, LocalFs, S3Bucket};
+use crate::{broker, SafeKeeperConf, wal_storage};
+
+
+use once_cell::sync::OnceCell;
+
+static BACKUP_RUNTIME: OnceCell<Runtime> = OnceCell::new();
+const DEFAULT_BACKUP_RUNTIME_SIZE: u32 = 16;
+
+const MIN_WAL_SEGMENT_SIZE: usize = 1 << 20; // 1MB
+const MAX_WAL_SEGMENT_SIZE: usize = 1 << 30; // 1GB
+const DEFAULT_WAL_SEGMENT_SIZE: usize = 16 << 20; // 16MB
+
+const BACKUP_ELECTION_PATH: &str = "WAL_BACKUP";
+
+async fn detect_task(
+    conf: SafeKeeperConf,
+    timeline_id: ZTenantTimelineId,
+    backup_start: Lsn,
+    mut wal_seg_size: Receiver<u32>,
+    lease_id : Option<i64>, 
+    mut lsn_durable: Receiver<Lsn>, 
+    lsn_backed_up: Sender<Lsn>,
+    ) -> Result<()> {
+
+    // TODO: antons come up with better naming
+    wal_seg_size.changed().await.expect("Failed to recieve wal segment size");
+    let wal_seg_size = *wal_seg_size.borrow();
+    let mut wal_seg_size = usize::try_from(wal_seg_size).with_context(|| format!("Couldn't convert {} to usize", wal_seg_size))?;
+    if !(MIN_WAL_SEGMENT_SIZE..=MAX_WAL_SEGMENT_SIZE).contains(&wal_seg_size) {
+        warn!("Invalid wal seg size provided, should be at least 1Mb per postgres, setting to default value of {}", DEFAULT_WAL_SEGMENT_SIZE);
+        wal_seg_size = DEFAULT_WAL_SEGMENT_SIZE; 
+    }
+
+    let mut backup_lsn = backup_start;
+
+    loop {
+        if let Some(lease_id) = lease_id {
+            broker::become_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.ok();
+        }
+
+        while lsn_durable.changed().await.is_ok() {
+            let commit_lsn = *lsn_durable.borrow();
+
+            if let Some(lease_id) = lease_id {
+                if broker::is_leader(lease_id, BACKUP_ELECTION_PATH.to_string(), &timeline_id, &conf).await.is_err() {
+                    info!("Leadership is lost for timeline {}", timeline_id);
+                    break;
+                }
+            }
+
+            if commit_lsn < backup_lsn {
+                bail!("Commit lsn {} moved back {}", commit_lsn, backup_lsn);
+            }
+
+            if backup_lsn.segment_number(wal_seg_size) == commit_lsn.segment_number(wal_seg_size) {
+                continue;
+            }
+
+            info!("Woken up for lsn {} committed, will back it up. backup lsn {}", commit_lsn, backup_lsn);
+
+            backup_lsn = backup_lsn_range(backup_lsn, commit_lsn, wal_seg_size, &conf, timeline_id).await?;
+            lsn_backed_up.send(backup_lsn)?;
+        }
+    }
+}
+
+pub async fn backup_lsn_range(start_lsn: Lsn, end_lsn: Lsn, wal_seg_size: usize, conf: &SafeKeeperConf, timeline_id: ZTenantTimelineId) -> Result<Lsn> {
+
+    let mut res = start_lsn;
+    for s in get_segments(start_lsn, end_lsn, wal_seg_size) {
```

Comment:
> .await

Not familiar with safekeeper internals, unfortunately, but how many such segments do we expect to receive?
Now we process them "one by one", with one task at a time waiting on `.await` and the rest of the sync tasks just queuing.

We can use [`FuturesUnordered`](https://github.com/neondatabase/neon/blob/81cad6277a2666ca47b97848628ffeafd6bf6aba/pageserver/src/remote_storage/storage_sync/upload.rs#L156) or its ordered counterpart to make it more concurrent.

Reply:
It's better to upload segments one by one, so we will have no holes in the uploaded range.

Reply:
at this one-by-one should be fine, agree that parallel upload will be a desirable functionality once larger-than-free tier, but I would bundle that with time based upload